python/lib/get_session_info.py

```diff
@@ -0,0 +1,328 @@
+
+def create_session(
+    env: Env,
+    candidate: DbCandidate,
+    create_session_info: CreateSessionInfo,
+    visit_label: str,
+    visit_number: int,
+) -> DbSession:
+    """
+    Create a session based on the parameters provided.
+    """
+
+    session = DbSession(
+        candidate_id  = candidate.id,
+        visit_label   = visit_label,
+        visit_number  = visit_number,
+        site_id       = create_session_info.site.id,
+        current_stage = 'Not Started',
+        scan_done     = True,
+        submitted     = False,
+        project_id    = create_session_info.project.id,
+        cohort_id     = create_session_info.cohort.id if create_session_info.cohort is not None else None,
+    )
+
+    env.db.add(session)
+    env.db.commit()
+
+    return session
+
+
+def get_session_site(env: Env, session_config: SessionConfig, site_id: int) -> DbSite:
+    """
+    Get the site for a session configuration site ID, or raise a `SessionConfigError` if that ID is
+    incorrect.
+    """
+
+    site = try_get_site_with_id(env.db, site_id)
+    if site is None:
+        raise_session_config_error(session_config, f"No site found for site ID {site_id}.")
+
+    return site
+
+
+def get_session_project(env: Env, session_config: SessionConfig, project_id: int) -> DbProject:
+    """
+    Get the project for a session configuration project ID, or raise a `SessionConfigError` if that
+    ID is incorrect.
+    """
+
+    project = try_get_project_with_id(env.db, project_id)
+    if project is None:
+        raise_session_config_error(session_config, f"No project found for project ID {project_id}.")
+
+    return project
+
+
+def get_session_cohort(env: Env, session_config: SessionConfig, cohort_id: int) -> DbCohort:
+    """
+    Get the cohort for a session configuration cohort ID, or raise a `SessionConfigError` if that
+    ID is incorrect.
+    """
+
+    cohort = try_get_cohort_with_id(env.db, cohort_id)
+    if cohort is None:
+        raise_session_config_error(session_config, f"No cohort found for cohort ID {cohort_id}.")
+
+    return cohort
+
+
+def get_candidate_next_visit_number(candidate: DbCandidate) -> int:
+    """
+    Get the next visit number for a new session for a given candidate.
+    """
+
+    visit_numbers = [
+        session.visit_number
+        for session in candidate.sessions
+        if session.visit_number is not None
+    ]
+
+    return max(*visit_numbers, 0) + 1
+
+
+def raise_session_config_error(session_config: SessionConfig, message: str) -> Never:
+    """
+    Raise a session configuration error for a given session configuration.
+    """
+
+    match session_config:
+        case SessionCandidateConfig():
+            session_name = f"{session_config.psc_id} / {session_config.cand_id} / {session_config.visit_label}"
+        case SessionPhantomConfig():
+            session_name = session_config.name
+
+    raise SessionConfigError(f"Error in session configuration ({session_name}):\n{message}")
```

Comment:
same, "session configuration" terminology could be simplified to point the user to check their configuration file.